src/Application/Commands/RedistributeMatchFunds.php

```diff
@@ -0,0 +1,87 @@
+<?php
+
+declare(strict_types=1);
+
+namespace MatchBot\Application\Commands;
+
+use MatchBot\Domain\CampaignFundingRepository;
+use MatchBot\Domain\DonationRepository;
+use MatchBot\Domain\FundingWithdrawal;
+use Symfony\Component\Console\Input\InputInterface;
+use Symfony\Component\Console\Output\OutputInterface;
+
+/**
+ * Redistribute match funding allocations where possible, from lower to higher priority match fund pots.
+ */
+class RedistributeMatchFunds extends LockingCommand
+{
+    protected static $defaultName = 'matchbot:redistribute-match-funds';
+
+    public function __construct(
+        private CampaignFundingRepository $campaignFundingRepository,
+        private DonationRepository $donationRepository
+    ) {
+        parent::__construct();
+    }
+
+    protected function configure(): void
+    {
+        $this->setDescription('Moves match funding allocations from lower to higher priority funds where possible');
+    }
+
+    protected function doExecute(InputInterface $input, OutputInterface $output): int
+    {
+        // TODO Change the fixed lookback to 2 days, or parameter-ise it, once CC23 is tidied in Prod.
+        $donationsToCheck = $this->donationRepository->findWithMatchingWhichCouldBeReplacedWithHigherPriorityAllocation(
+            new \DateTimeImmutable('-3 weeks')
+        );
+
+        $donationsAmended = 0;
+        foreach ($donationsToCheck as $donation) {
+            $highestAllocationOrderUsedForDonation = 0;
```

Comment:
AFAIK there's no rule that says ranks have to be positive, so when we haven't seen any we should start at INT_MIN. Or maybe clearer to start at null and give that special handling inside the loop.

```suggestion
            $highestAllocationOrderUsedForDonation = \PHP_INT_MIN;
```

Reply:
Alternatively `allocationOrder` could be a good candidate to turn into a value object.

Reply:
I agree longer term but had very limited time left before we said we'd patch CC data. I have made its setter throw on values < 0 for now instead.